### ISSUE TITLE
Exception during IEnumerable param enumeration pretends to be a query problem

### ISSUE THREAD

**Issue by Timovzl** [dotnet/efcore#30009]
When EF receives an `IEnumerable<T>` query parameter, and an exception happens to be thrown during its deferred enumeration, EF logs an error as if something went wrong with the query. In actuality, the error had nothing to do with the query.

Let's consider why one passes a deferred enumerable. At least not with the intent of making its enumeration a part of the query execution. 😃 Rather, say, because extra calls to `ToList()` reduce readability or are simply unnecessary.

For example:

```cs
var orderIds = inputModels.Select(model => model.OrderId);
// ToList() here would cause clutter - we are only enumerating once, after all

var orders = await dbContext.Orders
	.Where(order => orderIds.Contains(order.Id)
	.ToListAsync(cancellationToken);
```

Say that an exception is thrown during the enumeration of the `IEnumerable<T>`. A common use case for that is when we use DDD-style ValueObjects, such as an `OrderId` type. Perhaps `OrderId` is a string constrained to 50 alphanumeric characters. When the input does not satisfy the constraint, the `OrderId` constructor throws.

Suddenly, it matters way too much whether or not we add `ToList()` to the example's first line:

- If we add `ToList()`, the exception occurs normally. Our exception handler might kick in, blame the exception on caller error, and return BadRequest, without polluting the logs with any errors.
- If we omit `ToList()`, the same things happen, but not before EF has logged an error!

Irrelevant error entries are the bane of any log's usability, so the latter is quite undesirable. We could work around that by opting for the former, but how do we enforce that choice? Since the example was a natural way to write the code, spotting the "missing" `ToList()` would be an unreliable exercise.

## Suggested solution

Exceptions caught during the enumeration of any `IEnumerable` query parameter should _not_ cause EF to log. There is no difference of intent between the developer materializing the enumerable vs. letting EF enumerate it, so there should be no difference in the outcome either.

## Implementation ideas

Some options:

- Materialize any non-materialized enumerable parameters upfront, before entering the relevant try/catch block. (For example, if `is not IReadOnlyCollection<T>`, then we must materialize.)
- Enumerate any enumerable parameters in their own try/catch block. Wrap potential exceptions in a custom exception type. At the existing _outer_ catch that normally logs and rethrows, add a different catch for the new exception type. It simply unwraps the original exception and rethrows it.

### Provider and version information

EF Core version: 7.0.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 7.0
Operating system: Windows 11
IDE: Visual Studio 2022 17.4.3

**Comment by roji** [comment]
Can you post the exception that gets thrown?

**Comment by Timovzl** [comment]
Sure, let me post a minimal repro where I throw an arbitrary exception (I've chosen `InvalidDataException` here) and get the following error logged:

```
fail: Microsoft.EntityFrameworkCore.Query[10100]
      An exception occurred while iterating over the results of a query for context type 'TestApplication.Program+TestDbContext'.
```

## Minimal repro

```cs
public static class Program
{
	public static async Task Main()
	{
		var builder = WebApplication.CreateBuilder();

		builder.Services.AddSingleton<IHostedService, TestRunner>();
		builder.Services.AddDbContextFactory<TestDbContext>(options => options.UseSqlServer(
			@"Data Source=(LocalDB)\MSSQLLocalDB;Integrated Security=True;Initial Catalog=MultipleIncludeTest;"));

		using var app = builder.Build();

		await app.StartAsync();
		await app.StopAsync();
	}

	private class TestRunner : IHostedService
	{
		private IDbContextFactory<TestDbContext> DbContextFactory { get; }

		public TestRunner(IDbContextFactory<TestDbContext> dbContextFactory)
		{
			this.DbContextFactory = dbContextFactory;
		}

		public async Task StartAsync(CancellationToken cancellationToken)
		{
			await using var dbContext = await this.DbContextFactory.CreateDbContextAsync();

			await dbContext.Database.EnsureDeletedAsync();
			await dbContext.Database.EnsureCreatedAsync();

			var ids1 = EnumerateIdsSuccessfully();
			var ids2 = EnumerateIdsUnsuccessfully(); // Throws when enumerated

			// Pass an IEnumerable that is fine - no issues here
			_ = await dbContext.Parents
				.Where(parent => ids1.Contains(parent.Id))
				.ToListAsync();

			// Pass a dirty IEnumerable - an error gets logged here
			_ = await dbContext.Parents
				.Where(parent => ids2.Contains(parent.Id))
				.ToListAsync();
		}

		public Task StopAsync(CancellationToken cancellationToken)
		{
			return Task.CompletedTask;
		}

		private static IEnumerable<int> EnumerateIdsSuccessfully()
		{
			yield return 1;
			yield return 2;
		}

		private static IEnumerable<int> EnumerateIdsUnsuccessfully()
		{
			yield return 1;
			throw new InvalidDataException("Something went wrong.");
		}
	}

	public class TestDbContext : DbContext
	{
		public DbSet<Parent> Parents { get; private init; }

		public TestDbContext(DbContextOptions<TestDbContext> options)
			: base(options)
		{
		}
	}

	public class Parent
	{
		public int Id { get; set; }
	}
}
```

**Comment by Timovzl** [comment]
@roji Here is the exception's stack trace:

```
   at TestApplication.Program.TestRunner.<EnumerateIdsUnsuccessfully>d__7.MoveNext() in C:\Source\TestApplication\Program.cs:line 72
   at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.<VisitIn>g__ProcessInExpressionValues|31_0(SqlExpression valuesExpression, Boolean extractNullValues)
   at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.VisitIn(InExpression inExpression, Boolean allowOptimizedExpansion, Boolean& nullable)
   at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Visit(SqlExpression sqlExpression, Boolean allowOptimizedExpansion, Boolean preserveColumnNullabilityInformation, Boolean& nullable)
   at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Visit(SqlExpression sqlExpression, Boolean allowOptimizedExpansion, Boolean& nullable)
   at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Visit(SelectExpression selectExpression)
   at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Process(Expression queryExpression, IReadOnlyDictionary`2 parameterValues, Boolean& canCache)
   at Microsoft.EntityFrameworkCore.SqlServer.Query.Internal.SqlServerParameterBasedSqlProcessor.ProcessSqlNullability(Expression selectExpression, IReadOnlyDictionary`2 parametersValues, Boolean& canCache)
   at Microsoft.EntityFrameworkCore.Query.RelationalParameterBasedSqlProcessor.Optimize(Expression queryExpression, IReadOnlyDictionary`2 parametersValues, Boolean& canCache)
   at Microsoft.EntityFrameworkCore.SqlServer.Query.Internal.SqlServerParameterBasedSqlProcessor.Optimize(Expression queryExpression, IReadOnlyDictionary`2 parametersValues, Boolean& canCache)
   at Microsoft.EntityFrameworkCore.Query.Internal.RelationalCommandCache.GetRelationalCommandTemplate(IReadOnlyDictionary`2 parameters)
   at Microsoft.EntityFrameworkCore.Internal.RelationCommandCacheExtensions.RentAndPopulateRelationalCommand(RelationalCommandCache relationalCommandCache, RelationalQueryContext queryContext)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.<InitializeReaderAsync>d__21.MoveNext()
```

And here is the full log entry (where I'm _guessing_ the duplication is caused by the retrying execution strategy):

```
Microsoft.EntityFrameworkCore.Query[10100]
      An exception occurred while iterating over the results of a query for context type 'TestApplication.Program+TestDbContext'.
      System.IO.InvalidDataException: Something went wrong.
         at TestApplication.Program.TestRunner.EnumerateIdsUnsuccessfully()+MoveNext() in C:\Source\TestApplication\Program.cs:line 72
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.<VisitIn>g__ProcessInExpressionValues|31_0(SqlExpression valuesExpression, Boolean extractNullValues)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.VisitIn(InExpression inExpression, Boolean allowOptimizedExpansion, Boolean& nullable)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Visit(SqlExpression sqlExpression, Boolean allowOptimizedExpansion, Boolean preserveColumnNullabilityInformation, Boolean& nullable)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Visit(SqlExpression sqlExpression, Boolean allowOptimizedExpansion, Boolean& nullable)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Visit(SelectExpression selectExpression)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Process(Expression queryExpression, IReadOnlyDictionary`2 parameterValues, Boolean& canCache)
         at Microsoft.EntityFrameworkCore.SqlServer.Query.Internal.SqlServerParameterBasedSqlProcessor.ProcessSqlNullability(Expression selectExpression, IReadOnlyDictionary`2 parametersValues, Boolean& canCache)
         at Microsoft.EntityFrameworkCore.Query.RelationalParameterBasedSqlProcessor.Optimize(Expression queryExpression, IReadOnlyDictionary`2 parametersValues, Boolean& canCache)
         at Microsoft.EntityFrameworkCore.SqlServer.Query.Internal.SqlServerParameterBasedSqlProcessor.Optimize(Expression queryExpression, IReadOnlyDictionary`2 parametersValues, Boolean& canCache)
         at Microsoft.EntityFrameworkCore.Query.Internal.RelationalCommandCache.GetRelationalCommandTemplate(IReadOnlyDictionary`2 parameters)
         at Microsoft.EntityFrameworkCore.Internal.RelationCommandCacheExtensions.RentAndPopulateRelationalCommand(RelationalCommandCache relationalCommandCache, RelationalQueryContext queryContext)
         at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.InitializeReaderAsync(AsyncEnumerator enumerator, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
      System.IO.InvalidDataException: Something went wrong.
         at TestApplication.Program.TestRunner.EnumerateIdsUnsuccessfully()+MoveNext() in C:\Source\TestApplication\Program.cs:line 72
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.<VisitIn>g__ProcessInExpressionValues|31_0(SqlExpression valuesExpression, Boolean extractNullValues)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.VisitIn(InExpression inExpression, Boolean allowOptimizedExpansion, Boolean& nullable)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Visit(SqlExpression sqlExpression, Boolean allowOptimizedExpansion, Boolean preserveColumnNullabilityInformation, Boolean& nullable)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Visit(SqlExpression sqlExpression, Boolean allowOptimizedExpansion, Boolean& nullable)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Visit(SelectExpression selectExpression)
         at Microsoft.EntityFrameworkCore.Query.SqlNullabilityProcessor.Process(Expression queryExpression, IReadOnlyDictionary`2 parameterValues, Boolean& canCache)
         at Microsoft.EntityFrameworkCore.SqlServer.Query.Internal.SqlServerParameterBasedSqlProcessor.ProcessSqlNullability(Expression selectExpression, IReadOnlyDictionary`2 parametersValues, Boolean& canCache)
         at Microsoft.EntityFrameworkCore.Query.RelationalParameterBasedSqlProcessor.Optimize(Expression queryExpression, IReadOnlyDictionary`2 parametersValues, Boolean& canCache)
         at Microsoft.EntityFrameworkCore.SqlServer.Query.Internal.SqlServerParameterBasedSqlProcessor.Optimize(Expression queryExpression, IReadOnlyDictionary`2 parametersValues, Boolean& canCache)
         at Microsoft.EntityFrameworkCore.Query.Internal.RelationalCommandCache.GetRelationalCommandTemplate(IReadOnlyDictionary`2 parameters)
         at Microsoft.EntityFrameworkCore.Internal.RelationCommandCacheExtensions.RentAndPopulateRelationalCommand(RelationalCommandCache relationalCommandCache, RelationalQueryContext queryContext)
         at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.InitializeReaderAsync(AsyncEnumerator enumerator, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.ExecuteAsync[TState,TResult](TState state, Func`4 operation, Func`4 verifySucceeded, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
```

**Comment by roji** [comment]
@Timovzl if I'm understanding correctly, your exception ("something went wrong") appears in the exception stack trace - including the stack frame showing the enumeration (`EnumerateIdsUnsuccessfully()+MoveNext()`), so it seems pretty clear where the error is coming from. Are you specifically asking for the wording of the exception ("... while iterating over the results of a query ...") to be changed?

**Comment by roji** [comment]
Well the EF query fails to execute (because of an application-thrown exception)... I can see how maybe the log message could be different for that case, but I do think any EF query failure should be logged regardless of the source of the issue...

**Comment by Timovzl** [comment]
Oh, the exception itself is entirely correct and expected. No problems there.

Indeed, I'm asking for no error to be logged. An informational-level would be totally fine, but anything from warning and up is disruptive.

I have tried to substantiate this as clearly as I could in the original post. If we materialize the list upfront, we simply get the exception and full control over how to handle it. By contrast, if we pass an _unmaterialized_ `IEnumerable` and the exception happens during query construction, EF _logs an error_ on top, even though we have yet to decide if the exception was error-worthy in the first place!

The workaround is to always pre-materialize anything that is passed as a parameter, but as shown in the example, that is a hard agreement to guard, since the code is perfectly natural when passing the (unmaterialized) result of `.Select()`.

> I do think any EF query failure should be logged regardless of the source of the issue...

I certainly understand that sentiment in general. But does it really cover this scenario in a sensible way? Consider this: Even though the `IEnumerable` is only ever enumerated once in either scenario, the behavior changes significantly if we call `ToList()` manually vs. if we don't. That is hard to work with, since it's so unexpected.

Logging the error takes away some control from the user of how to handle the exception: if we return BadRequest and the caller is at fault, then error logs (and alerts triggering based on them) are false alarms.

Perhaps we might look at it this way: Any EF query failure should be logged, but materialization failure of a user `IEnumerable` has nothing to do with EF and is practically something that happened _just before_ the EF query.

**Comment by roji** [comment]
> Even though the IEnumerable is only ever enumerated once in either scenario, the behavior changes significantly if we call ToList() manually vs. if we don't. That is hard to work with, since it's so unexpected.

But that's always going to be the case, since in one case the exception happens from inside EF's stack trace, and in the other case it doesn't. For example, even if we stop logging for this case, the actual exception that gets thrown will be different - in one case it is thrown from ToList over the enumerable, in the other from EF. The types will be different.

> Logging the error takes away some control from the user of how to handle the exception: if we return BadRequest and the caller is at fault, then error logs (and alerts triggering based on them) are false alarms.

How is control taken away? If I'm following correctly, we're only discussing the log message, and agree that the exception should be thrown from EF just as it is today?

In any case, in my mind the log message really only indicates that an EF query failed; this can happen because of network issues, some problem in the query itself, or because the data provided by the user is somehow problematic (as in this case). The basic requirement here is to be able to look at the logs and see queries that failed, regardless of the reason, IMHO (that's available by inspecting the exception details and stack trace).

> [...] materialization failure of a user IEnumerable has nothing to do with EF and is practically something that happened just before the EF query.

The fact that you've handed something with deferred evaluation (i.e. an IEnumerable) to EF doesn't mean that its evaluated doesn't happen as part of the query... As above, I suspect there are other APIs out there which accept IEnumerable, and the effect of enumeration failing is going to look different in various ways then if you yourself called ToList on it and passed the results instead.

Stepping back, note that there are various other cases where evaluating something inside a query or outside can lead to behavioral changes. For example, the following two queries have very different behaviors:

```c#
var blogs = ctx.Blogs.Where(b => b.Name == "Foo").ToList();

var name = "Foo";
var blogs = ctx.Blogs.Where(b => b.Name == name).ToList();
```

This isn't exactly the same as your scenario, but it's also a case where doing something outside of the query vs. inside is going to change how things work. I'm guessing that if I think about it some more I could find other examples which may be more similar.

**Comment by Timovzl** [comment]
> If I'm following correctly, we're only discussing the log message, and agree that the exception should be thrown from EF just as it is today?

Agreed.

> How is control taken away?

In my mind, a properly maintainable application only logs warnings and errors when something unexpected is the matter and investigation is warranted. Generally, this reveals a bug and requires a fix. As for transient errors, we have good tooling for retrying automatically. But what I believe should definitely be avoided is for an external API caller to have the power to spam the team's error logs. For caller error, we have the 4xx range of HTTP responses, a clear indication that our end is functioning as expected. We do not want to disturb the development team every time a caller misinterprets or misuses our API.

We also cannot simply suppress EF query execution error logs, as those may very well reveal bugs in other cases.

The issue I'm running into is that EF is sounding the alarms. The control that I seek to maintain is how exceptions thrown by our own code are handled, e.g. only returning BadRequest and maybe logging an informational entry.

> the actual exception that gets thrown will be different - in one case it is thrown from ToList over the enumerable, in the other from EF. The types will be different.

I see only the stack traces differing. Those are normally used for debugging purposes. Error handling uses the type and/or contents of the exception, which are unchanged. The stack trace will always differ if we enumerate in a deferred fashion. I think that's expected and unproblematic.

> For example, the following two queries have very different behaviors

I agree that "doing something inside a query vs. outside of it" can lead to different results.

But is the following distinction a case of "doing something inside a query"? I actually think not.

```cs
dbContext.Blogs.Where(b => myEnumerable.Contains(b.Id));
dbContext.Blogs.Where(b => myList.Contains(b.Id));
```

What do we know must happen here? A query must be written, with parameters obtained from a sequence of .NET objects. To this end, the sequence must be enumerated. In both cases, the sequence's objects are enumerated before we can complete the writing of the query, and the queries will look identical.

As for differences, the deferred `IEnumerable` might throw during enumeration. Purely from the perspective of query generation, it seems to me there's no clear answer as to whether or not we should log an error in that case. Perhaps we might mildly lean towards always logging an error regardless of the reason, for simplicity.

**But I think a much more solid basis for the decision is the developer's intent.**

Could you think of a reason that the caller would say: "I'm deliberately _not_ enumerating this `IEnumerable` yet, so that EF will log an error for me if my own code throws - even though I'm also getting an exception that I could catch." Would you agree that this reasoning is fictional and unrealistic?

More realistically, the intent of passing a deferred `IEnumerable` is to optimize for readability and to avoid needless materialization. A simple rule of thumb followed by many is that you do not call `ToList()` unless you have a clear reason to. Normally that reason is multiple enumeration. (ReSharper even has an analyzer for this.)

If you can agree with me that that is why people pass a deferred `IEnumerable`, then it follows that behavioral differences are not aligned with the intent.

I would consider working around the issue, but it is not easily visible that the deferred `IEnumerable` could cause unwanted behavior! Whereas the multiple enumeration scenario can be fairly easily observed ("if you want to enumerate any given `IEnumerable` more than once, instead materialize it"), there is no clear indicator here. The code looks perfectly fine. I don't think we can rely on people catching this mistake in the code review every time.

**Comment by roji** [comment]
> But what I believe should definitely be avoided is for an external API caller to have the power to spam the team's error logs.

There's nothing about an IEnumerable that implies it's coming from an external API caller. That may sometimes be the case, but in other cases (I'd venture to say most) the IEnumerable is constructed inside your own application. It seems you're equating IEnumerable with "provided externally over HTTP", which certainly isn't the only (or even frequent) case.

> But is the following distinction a case of "doing something inside a query"? I actually think not.

Absolutely - providing a List or an IEnumerable to an API is a very different thing. At minimum, that API may enumerate your IEnumerable multiple times, which apart from causing perf issues could even have side effects. The expectation that an API behave in an identical way given an IEnumerable or a List simply doesn't work.

As to why people pass IEnumerable, I think there's a bit of speculation there, and in any case it seems a bit orthogonal to how EF should log on failure. In other words, the fact someone passes an IEnumerable just for readability (to avoid doing ToList) doesn't necessarily mean they're not interesting in an EF log message if the query using that IEnumerable fails because of an exception.

> Whereas the multiple enumeration scenario can be fairly easily observed ("if you want to enumerate any given IEnumerable more than once, instead materialize it"), there is no clear indicator here.

The two seem equivalent to me. Lots of developer mistakenly produce multiple enumeration by passing an IEnumerable - I don't think it's easily observed at all. There's nothing on an API which tells you whether it's going to enumerate your IEnumerable argument more than once.

> The code looks perfectly fine. I don't think we can rely on people catching this mistake in the code review every time.

I'm not following... How would not logging from EF help anyone catch anything here? Which mistake are you referring to exactly?

In any case, I think the arguments on either side are pretty clear; we'll discuss this as a team and see what everyone thinks about it.

**Comment by ajcvickers** [comment]
Note from triage: we discussed this, but we don't plan to make any changes here. If an exception is thrown while EF is processing a query than it will be logged regardless of the source of that exception.

**Comment by Timovzl** [comment]
@roji Thanks for the extensive discussion!

I will look for a workaround. Perhaps you can offer suggestions. Here are my ideas:

- Is there an interceptor that would let me pre-materialize any plain `IEnumerable<T>` just-in-time, when a query is about to be written?
- Is there a way I could filter out these log messages without filtering _all_ query-related log messages?
- I could simply suppress _query_ error logging altogether, since I'm handling all exceptions anyway, logging the errors (where needed). Doing so should render EF's query error logging nonessential, or am I overlooking something?

As a side note, in case you'd like to act on it, the message is not entirely correct: "An exception occurred while iterating over the results of a query". The query is not fully written yet, let alone executed and iterating its results.

I response to your last two comments, for the sake of completion:

> I'm not following... How would not logging from EF help anyone catch anything here? Which mistake are you referring to exactly?

I was referring to the workaround of always manually calling `ToList()` before passing any `IEnumerable<T>` to EF. That would ensure that the enumeration exceptions happen before EF is touching the enumerables, and thus avoid the error logging issue.

The mistake I was referring to is forgetting to manually call `ToList()` on an `IEnumerable<T>` passed to EF, i.e. forgetting to apply the workaround. That'd be pretty hard to consistently spot in code reviews, making this particular workaround... unreliable.

> There's nothing on an API which tells you whether it's going to enumerate your IEnumerable argument more than once.

Getting off-topic, but... it really, _really_ shouldn't. Any method that asks for `IEnumerable<T>` _should_ only enumerate the input once. If it wants to enumerate multiple times, it should choose to either materialize the input once and work with the result (`input as IReadOnlyCollection<T> ?? input.ToList()`), or asks for `IReadOnlyCollection<T>` to demand a materialized collection.

I believe that the framework libraries adhere strictly to this convention. It's hard to blame the caller if you wanted a collection but asked for any sequence. :smile:

**Comment by roji** [comment]
> Perhaps you can offer suggestions [...]

Nothing comes to mind unfortunately; this particular enumeration happens for each query execution (i.e. it's not part of the once-only compilation process), so there's no interceptor for that. You can of course create a wrapper of your own for querying which would visit the query tree and do the enumeration, but that's a very heavy hammer (and not great for perf, which is one reason we don't have an interceptor here in the first place).

> Is there a way I could filter out these log messages without filtering all query-related log messages?

If this is a specific exception which you can identify in a stack trace, you could have a logging filter that does this. But this doesn't work in the general case where any exception may be thrown.

> Any method that asks for IEnumerable<T> should only enumerate the input once.

That should be true in most cases, though there are exceptions. But in any case there's no hard, visible guarantee around that.

I have to say I'm curious as to why this log message is bothersome enough for you to attempt such complex things to try to avoid it... It really doesn't seem that bad to me personally.

**Comment by Timovzl** [comment]
Yeah, I got _really_ close with [one of the new](https://devblogs.microsoft.com/dotnet/announcing-ef7-preview7-entity-framework/#linq-expression-tree-interception) interceptors, [IQueryExpressionInterceptor](https://github.com/dotnet/efcore/blob/main/src/EFCore/Diagnostics/IQueryExpressionInterceptor.cs), but unfortunately it is called a fraction too late: _after_ any captured variable expressions (like our `IEnumerable<T>`) have already been replaced by placeholders, e.g. `__ids_0`. The variable values themselves never pass through that interceptor, unfortunately.

> I have to say I'm curious as to why this log message is bothersome enough for you to attempt such complex things to try to avoid it... It really doesn't seem that bad to me personally.

Fair question. With regards to warnings and errors, we've worked hard to keep our logs clean of anything that does not require attention. In my experience in a consultant role, distractions like that accumulate and become velocity killers, so I'm pretty keen on optimizing that part of the workflow. I feel strongly about the scenario we've been discussing because it's triggered by external API callers. I really want to prevent API callers from (inadvertently or maliciously) straining the development team. It truly works to have no boy who cries wolf!
